medcat/rel_cat.py

```diff
@@ -0,0 +1,643 @@
+
+            for i, data in enumerate(train_dataloader, 0):
+                self.model.train()
+                # self.model.zero_grad()
+
+                current_batch_size = len(data[0])
+                token_ids, e1_e2_start, labels, _, _, _ = data
+
+                attention_mask = (
+                    token_ids != self.pad_id).float().to(self.device)
+                token_type_ids = torch.zeros(
+                    (token_ids.shape[0], token_ids.shape[1])).long().to(self.device)
+                labels = labels.to(self.device)
+
+                model_output, classification_logits = self.model(
+                    input_ids=token_ids,
+                    token_type_ids=token_type_ids,
+                    attention_mask=attention_mask,
+                    e1_e2_start=e1_e2_start
+                )
+
+                batch_loss = criterion(
+                    classification_logits.view(-1, self.nclasses).to(self.device), labels.squeeze(1))
+                batch_loss.backward()
+                batch_loss = batch_loss / gradient_acc_steps
+
+                total_loss += batch_loss.item() / current_batch_size
+
+                batch_acc, _, batch_precision, batch_f1, _, _, batch_stats_per_label = self.evaluate_(
+                    classification_logits, labels, ignore_idx=-1)
+
+                loss_per_batch.append(batch_loss / current_batch_size)
+                accuracy_per_batch.append(batch_acc)
+
+                torch.nn.utils.clip_grad_norm_(
+                    self.model.parameters(), max_grad_norm)
+
+                if (i % gradient_acc_steps) == 0:
+                    self.optimizer.step()
+                    self.scheduler.step()
+
+                if ((i + 1) % current_batch_size == 0):
+                    self.log.debug(
+                        "[Epoch: %d, loss per batch, accuracy per batch: %.3f, %.3f, average total loss %.3f , total loss %.3f]" %
+                        (epoch, loss_per_batch[-1], accuracy_per_batch[-1], total_loss / (i + 1), total_loss))
+
+                pbar.update(current_batch_size)
+
+            pbar.close()
+
+            if len(loss_per_batch) > 0:
+                losses_per_epoch.append(
+                    sum(loss_per_batch) / len(loss_per_batch))
+                self.log.info("Losses at Epoch %d: %.5f" %
+                              (epoch, losses_per_epoch[-1]))
+
+            if len(accuracy_per_batch) > 0:
+                accuracy_per_epoch.append(
+                    sum(accuracy_per_batch) / len(accuracy_per_batch))
+                self.log.info("Train accuracy at Epoch %d: %.5f" %
+                              (epoch, accuracy_per_epoch[-1]))
+
+            total_loss = total_loss / (i + 1)
+
+            end_time = datetime.now().time()
+
+            print(
+                "======================== TRAIN SET TEST RESULTS ========================")
+            _ = self.evaluate_results(train_dataloader, self.pad_id)
+
+            print(

```

Comment:
I know some of CAT stats are also still under `print` rather than a logger. But is there a reason for it to be the case here?